src/file.ts

```diff
@@ -3608,28 +3614,60 @@ class File extends ServiceObject<File> {
     }
     const returnValue = retry(
       async (bail: (err: Error) => void) => {
-        await new Promise<void>((resolve, reject) => {
+        if (data instanceof Readable && !data.readable) {
+          // Readable streams can only be consumed once and would cause a dead
+          // lock when used with pipeline().
+          return bail(new Error(FileExceptionMessages.STREAM_NOT_READABLE));
+        }
```

Comment:
Is this true? I've tested this locally and pipeline seems to exit cleanly. Example:

```ts
const readable = stream.Readable({read() {this.push(null)}});
readable.read();
readable.readable // false
stream.pipeline(a, stream.PassThrough, () => { console.log('done')})
```

Reply:
When I run your code on Node `v16.20.1`, `done` is not printed. This means that the pipeline callback is never called. If you were to resolve/reject a promise in said callback, you would have a forever hanging promise.

```js
const stream = require('stream');

const readable = stream.Readable({read() {this.push(null)}});
readable.read();
readable.readable // false
stream.pipeline(readable, stream.PassThrough, () => { console.log('Never called!!!')})
```

Reply:
Consider the following example:
```js
const stream = require('stream');

const readable = stream.Readable({
  read() {
    this.push('aaa');
    this.push(null);
  },
});
readable.read();
readable.readable; // false
stream.pipeline(
  readable,
  async function* (src) {
    for await (const chunk of src) {
      console.log(chunk);
    }
  },
  err => {
    console.log(err ? 'Pipeline failed.' : 'Pipeline succeeded.');
  }
);
```

Running this results in "Pipeline succeeded" being printed while no chunk was actually "piped". I adapted the comment in my commit to better reflect what happens.

Reply:
Note that maybe this is an expected behavior in NodeJS: If a stream is piped twice, the second time it will appear "empty" (as if it contained 0 bytes). If that is the case, maybe this `if` should be removed ? It's kind of an edge case but on the other hand, as a dev, if I mistakenly `file.save(alreadyConsumed)` I would rather see an error than upload an empty file...

Reply:
I reported an [issue](https://github.com/nodejs/node/issues/48764) in NodeJS regarding this.

Reply:
I also added an await for next tick to better detect wether the stream is not readable:

```js
const stream = require('stream');
const {pipeline} = require('stream/promises');

async function main() {
  const readable = stream.Readable({
    read() {
      this.push('aaa');
      this.push(null);
    },
  });

  console.log(readable.readable); // true

  readable.read();

  console.log(readable.readable); // true
  console.log(readable.destroyed); // false
  console.log(readable._readableState.ended); // true

  await new Promise((resolve) => setImmediate(resolve));

  console.log(readable.readable); // false
  console.log(readable.destroyed); // true
  console.log(readable._readableState.ended); // true

  await pipeline(readable, async function* (source) {
    for await (const chunk of source) {
      console.log('chunk', chunk);
    }
    console.log('done');
  });
}

main();

```